File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {	
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		int length=0, A;
		int[] LIS;
		
		int N=Integer.parseInt(br.readLine());
		LIS=new int[N];
		
		StringTokenizer st=new StringTokenizer(br.readLine());
		for(int i=0;i<N;i++) {
			int left=0, right=length, mid;
			LIS[i]=1<<25;
			A=Integer.parseInt(st.nextToken());
			
			while(left<right) {
				mid=(left+right)/2;
				if(LIS[mid]>A)
					right=mid-1;
				else if(LIS[mid]<A)
					left=mid+1;
				else
					break;
			}
			mid=(left+right)/2;
			
			if(A<=LIS[mid])
				LIS[mid]=A;
			else
				LIS[mid+1]=A;
			if(mid==length)
				length++;
		}
		
		System.out.println(length);
	}
}

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.c
#include<stdio.h>
#include<malloc.h>

int main(void)
{
	int length=0, N, A, *LIS=NULL;

	scanf("%d", &N);
	LIS=(int *)malloc(N*sizeof(int));

	for(int i=0;i<N;++i)
	{
		int left=0, right=length, mid;
		LIS[i]=10000000;
		scanf("%d", &A);

		while(left<right)
		{
			mid=left+right>>1;
			if(LIS[mid]>A)
				right=mid-1;
			else if(LIS[mid]<A)
				left=mid+1;
			else
				break;
		}
		mid=left+right>>1;

		if(A<=LIS[mid])
			LIS[mid]=A;
		else
			LIS[mid+1]=A;
		length+=mid==length;
	}

	printf("%d", length);
	free(LIS);
	return 0;
}